data-import/desktop-csv-import/desktop-csv-import.adoc

```diff
@@ -0,0 +1,420 @@
+
+Now that we know the types of nodes and relationships we will have and the properties involved, we can construct the Cypher statements to create the data for this model.
+
+=== Avoiding duplicates and increasing performance
+
+One final thing we need to think about before we create data in the graph is ensuring values are unique and performance is efficient.
+To handle this, we can use constraints.
+Just as with other databases, constraints ensure data integrity criteria is not violated, while simultaneously indexing the property with the constraint for faster query performance.
+
+In this exercise, we will add two constraints - one for `productId` and one for `orderId`.
+This will ensure when we create a new node of each of those types or a relationship to connect them, we know the entities are unique and indexed.
+
+Good practice is to add any constraints and indexes before loading to avoid existing violations and ensure data is added to the index as it is created.
+Below is the Cypher for adding indexes.
+
+[source, cypher]
+----
+CREATE CONSTRAINT ON (p:Product) ASSERT p.id IS UNIQUE;
+CREATE CONSTRAINT ON (o:Order) ASSERT o.id IS UNIQUE;
+----
+
+=== Cypher
+
+Now we are ready to create the Cypher for creating the data in the graph!
+We will use `CREATE` statements where we are sure that we will not have duplicate rows in our CSV file.
+For values we know exist in the database already, we will first use `MATCH` to find them before executing any updates or inserts.
+
+.Products
+To start, let us load the products into the graph.
+We start with our `LOAD CSV` statement from above, then we add the Cypher to create the data from the CSV into our graph model.
+We can use the `CREATE` statement, as we feel confident that we have our constraint in place, no duplicate rows in our file, and no existing data in our graph.
+The properties will be set to the converted values we handled earlier in this guide.
+
+[source, cypher]
+----
+LOAD CSV FROM 'file:///desktop-csv-import/products.csv' AS row
+CREATE (p:Product {productId: toInteger(row[0])})
+  SET p.productName = row[1], p.unitCost = toFloat(row[2])
+RETURN count(p)
+----
+
+If you run that statement, it will return the number of product nodes (`count(p)`) that were created in the database.
+You can cross-check that number with the number of rows in the CSV file from earlier (77 rows in products.csv).
+You can also run a validation query to return a sample of nodes and review that the properties look accurate.
+
+[source, cypher]
+----
+//validate products loaded correctly
+MATCH (p:Product)
+RETURN p LIMIT 20
+----
+
+image::{img}desktop_import_verify_products.jpg[link="{img}desktop_import_verify_products.jpg",role="popup-link"]
+
+.Orders
+Next, we can load the orders.
+Again, since we feel confident our constraint will catch violations and our CSV file data has no duplicates, we can use the `CREATE` statement to insert the data more efficiently.
```

Comment:
would be a good opportunity to show merge.

Reply:
also for idempotent runs.

Reply:
Done. Thanks for suggesting this!